File: ios/Classes/FlutterAudioRecorder4Plugin.swift
import Flutter
import UIKit
import AVFoundation

public class FlutterAudioRecorder4Plugin: PermissionRequestListenerActivityPlugin, AVAudioRecorderDelegate {

    private static let DEFAULT_SAMPLE_RATE_HZ = 16000
    private static let DEFAULT_PEAK_POWER = -120.0
    private static let DEFAULT_AVERAGE_POWER = -120.0
    private static let DEFAULT_DATA_SIZE_BYTES = 0
    private static let DEFAULT_BUFFER_SIZE_BYTES = 1024
    private static let DEFAULT_METERING_ENABLED = true
    private static let DEFAULT_RECORDER_BPP: UInt8 = 16
    private static let DEFAULT_IOS_AUDIO_CHANNEL = 0
    private static let IOS_POWER_LEVEL_FACTOR = 0.25// iOS factor : to match iOS power level
    
    
    private var sampleRateHz = DEFAULT_SAMPLE_RATE_HZ
    private var dataSizeBytes = DEFAULT_DATA_SIZE_BYTES
    private var peakPower = DEFAULT_PEAK_POWER
    private var averagePower = DEFAULT_AVERAGE_POWER
    private var recorderState = RecorderState.UNSET
    private var bufferSizeBytes = DEFAULT_BUFFER_SIZE_BYTES
    private var recorder: AVAudioRecorder?
    private var filepath: String?
    private var ext: String?
    private var meteringEnabled = DEFAULT_METERING_ENABLED
    private var message: String?
    private var iosAudioChannel: Int = DEFAULT_IOS_AUDIO_CHANNEL
    
    private var filepathTemp: String? {
        get {
            return filepath?.plus(".temp")
        }
    }
    
    private var duration: Int = 0
    
    private var recording: [String : Any?] {
        get {
            [
                NamedArguments.FILEPATH: filepath,
                NamedArguments.FILEPATH_TEMP: [filepathTemp, "bla"].compactMap { $0 }.joined(separator:""),
                NamedArguments.EXTENSION: ext,
                NamedArguments.DURATION: duration,
                NamedArguments.AUDIO_FORMAT: ext?.toAudioFormat()?.name,
                NamedArguments.RECORDER_STATE: recorderState.value,
                NamedArguments.METERING_ENABLED: meteringEnabled,
                NamedArguments.PEAK_POWER: peakPower,
                NamedArguments.AVERAGE_POWER: averagePower,
                NamedArguments.SAMPLE_RATE_HZ: sampleRateHz,
                NamedArguments.MESSAGE: message
            ]
        }
    }
    
    private var doProcessAudioStream = false
       
    var startTime: Date!
  
   
    override public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        var methodCall = call.method.toMethodCall()
        
        switch call.method.toMethodCall() {
        case .CURRENT:
            handleCurrent(result)
            break
        case .INIT:
            handleInit(call, result)
            break
        case .START:
            handleStart(result)
            break
        case .STOP:
            handleStop(result)
            break
        case .PAUSE:
            handlePause(result)
            break
        case .RESUME:
            handleResume(result)
            break
        default:
            super.handle(call, result: result)
            break
        }
    }
    
    
    private func handleInit(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if (recorderState == RecorderState.UNSET || recorderState == RecorderState.INITIALIZED || recorderState == RecorderState.STOPPED) {
            resetRecorder()
            
            let dic = call.arguments as! [String : Any]
            
            filepath = dic[NamedArguments.FILEPATH] as? String
            if (filepath == "") {
                var startTime = Date()
                let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
                filepath = documentsPath + "/" + String(Int(startTime.timeIntervalSince1970)) + ".m4a"
                print("path: " + (filepath ?? ""))
            }
            
            ext = dic[NamedArguments.EXTENSION] as? String
            sampleRateHz = dic[NamedArguments.SAMPLE_RATE_HZ] as? Int ?? sampleRateHz
            recorderState = filepath.isNotNullOrEmpty() && ext.isNotNullOrEmpty() ? RecorderState.INITIALIZED : RecorderState.UNSET
            message = "Recorder initialized"
            iosAudioChannel = dic[NamedArguments.IOS_AUDIO_CHANNEL] as? Int ?? FlutterAudioRecorder4Plugin.DEFAULT_IOS_AUDIO_CHANNEL
            result(recording)
        } else {
            result(FlutterError(code: "", message: "Recorder not re-initialized", details: "RecorderState is not UNSET, INITIALIZED, or STOPPED - i.e. currectly recording"))
        }
    }
        
    private func handleCurrent(_ result: @escaping FlutterResult) {
        result(recording)
    }
    
    private func handleStart(_ result: @escaping FlutterResult) {
        do {
            
            var audioSession = AVAudioSession.sharedInstance()
            
            #if swift(>=4.2)
            try audioSession.setCategory(AVAudioSession.Category.playAndRecord, options: AVAudioSession.CategoryOptions.defaultToSpeaker)
            #else
            try audioSession.setCategory(AVAudioSession.CategoryPlayAndRecord, with: AVAudioSessionCategoryOptions.defaultToSpeaker)
            #endif
            
            try audioSession.setActive(true)
            
            if let urlFilepath = filepath {
                
                var settings = [
                    AVFormatIDKey: ext?.toAudioExtension()?.audioFormatIdentifier,
                    AVSampleRateKey: sampleRateHz,
                    AVNumberOfChannelsKey: 1,
                    AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
                ]
                                
                recorder = try AVAudioRecorder(url: URL(string: urlFilepath)!, settings: settings)
                recorder?.delegate = self
                recorder?.isMeteringEnabled = (recording[NamedArguments.METERING_ENABLED] as? Bool) ?? FlutterAudioRecorder4Plugin.DEFAULT_METERING_ENABLED
                recorder?.prepareToRecord()
               
                startProcessing()
                result(recording)
            } else {
                result(FlutterError(code: "", message: "Recorder not started", details: "Problem with filepath"))
            }
        } catch {
            result(FlutterError(code: "", message: "Recorder not started", details: error))
        }
    }
    
    private func handlePause(_ result: @escaping FlutterResult) {
        recorderState = RecorderState.PAUSED
        stopProcessing(doRelase: false)
        resetPowers()
        result(recording)
    }
    
    private func handleResume(_ result: @escaping FlutterResult) {
        startProcessing()
        result(recording)
    }
    
    private func handleStop(_ result: @escaping FlutterResult) {
        if recorderState == RecorderState.STOPPED {
            result(recording)
        } else {
            recorderState = RecorderState.STOPPED
            stopProcessing(doRelase: true)
        
            result(recording)
        }
    }

    private func resetRecorder() {
        resetPowers()
        dataSizeBytes = FlutterAudioRecorder4Plugin.DEFAULT_DATA_SIZE_BYTES
    }
    
    private func startProcessing() {
        recorderState = RecorderState.RECORDING
        recorder?.record()
        doProcessAudioStream = true
        //TODO - CHRIS - recordingThread = Thread({ processAudioStream() }, "Audio Processing Thread")
        //TODO - CHRIS - recordingThread?.start()
    }
    
    private func stopProcessing(doRelase: Bool) {
        if (doRelase) {
            recorder?.stop()
            recorder = nil
        } else {
            recorder?.pause()
        }
        doProcessAudioStream = false
        
        //TODO - CHRIS - recordingThread = null
    }
    
    private func processAudioStream() {
        if let rec = recorder {
            updatePowers(rec)
            duration = Int(rec.currentTime * 1000)
            meteringEnabled = rec.isMeteringEnabled
        }
    }
    
    private func updatePowers(_ recorder: AVAudioRecorder) {
        recorder.updateMeters()
        peakPower = Double(recorder.peakPower(forChannel: iosAudioChannel))
        averagePower = Double(recorder.averagePower(forChannel: iosAudioChannel))
    }
    
    private func resetPowers() {
        peakPower = FlutterAudioRecorder4Plugin.DEFAULT_PEAK_POWER
        averagePower = FlutterAudioRecorder4Plugin.DEFAULT_AVERAGE_POWER
    }
}

File: android/src/main/kotlin/com/tcubedstudios/flutter_audio_recorder4/FlutterAudioRecorder4Plugin.kt
package com.tcubedstudios.flutter_audio_recorder4

import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.AUDIO_FORMAT
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.AVERAGE_POWER
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.DURATION
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.EXTENSION
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.FILEPATH
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.METERING_ENABLED
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.PEAK_POWER
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.RECORDER_STATE
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.SAMPLE_RATE
import android.Manifest.permission.RECORD_AUDIO
import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.media.AudioFormat.CHANNEL_IN_MONO
import android.media.AudioFormat.ENCODING_PCM_16BIT
import android.media.AudioRecord
import android.media.MediaRecorder.AudioSource.MIC
import android.os.Build.VERSION_CODES
import com.tcubedstudios.flutter_audio_recorder4.AudioExtension.Companion.toAudioFormat
import com.tcubedstudios.flutter_audio_recorder4.MethodCalls.*
import com.tcubedstudios.flutter_audio_recorder4.MethodCalls.Companion.toMethodCall
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.FILEPATH_TEMP
import com.tcubedstudios.flutter_audio_recorder4.NamedArguments.MESSAGE
import com.tcubedstudios.flutter_audio_recorder4.RecorderState.*
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import kotlin.math.abs
import kotlin.math.ln

//TODO - CHRIS - use a state machine
//https://github.com/Tinder/StateMachine -> supports kotlin and swift

//Registrar is passed for Android plugin v1 compatibility
class FlutterAudioRecorder4Plugin(
    registrar: Registrar? = null,
    methodChannel: MethodChannel? = null
) : PermissionRequestListenerActivityPlugin(registrar, methodChannel) {

  //TODO - CHRIS
  // If recorder is ever expected to operate in a background service, implement ServiceAware interface.

  companion object {
    private const val DEFAULT_PEAK_POWER = -120.0
    private const val DEFAULT_AVERAGE_POWER = -120.0
    private const val DEFAULT_DATA_SIZE = 0L
    private const val DEFAULT_BUFFER_SIZE = 1024
    private const val DEFAULT_METERING_ENABLED = true
    private const val RECORDER_BPP: Byte = 16;
    private const val IOS_POWER_LEVEL_FACTOR = 0.25// iOS factor : to match iOS power level
  }

  override val permissionsToRequest = listOf(
    PermissionToRequest(RECORD_AUDIO),
    PermissionToRequest(WRITE_EXTERNAL_STORAGE, maxSdk = VERSION_CODES.LOLLIPOP_MR1)//WR_EX_ST was removed in M, so last allowed version is LP_MR1
  )

  private var sampleRate = 16000L//Khz
  private var dataSize = DEFAULT_DATA_SIZE
  private var peakPower = DEFAULT_PEAK_POWER
  private var averagePower = DEFAULT_AVERAGE_POWER
  private var recorderState = UNSET
  private var bufferSize = DEFAULT_BUFFER_SIZE
  private var fileOutputStream: FileOutputStream? = null
  private var recordingThread: Thread? = null
  private var recorder: AudioRecord? = null
  private var filepath: String? = null
  private var extension: String? = null
  private var meteringEnabled = DEFAULT_METERING_ENABLED
  private var message: String? = null

  private val filepathTemp: String?
    get() = filepath?.plus(".temp")
  private val duration: Int
    get() = (dataSize / (sampleRate * 2 * 1)).toInt()
  private val recording: Map<String, Any?>
    get() = mapOf(
      FILEPATH to filepath,
      FILEPATH_TEMP to filepathTemp,
      EXTENSION to extension,
      DURATION to duration * 1000,
      AUDIO_FORMAT to extension?.toAudioFormat()?.name,
      RECORDER_STATE to recorderState.value,
      METERING_ENABLED to meteringEnabled,
      PEAK_POWER to peakPower,
      AVERAGE_POWER to averagePower,
      SAMPLE_RATE to sampleRate,
      MESSAGE to message
    )

  //region Handle method calls from flutter
  override fun onMethodCall(call: MethodCall, result: Result) {
    super.onMethodCall(call, result)

    when(call.method.toMethodCall()) {
      INIT -> handleInit(call, result)
      CURRENT -> handleCurrent(result)
      START -> handleStart(result)
      PAUSE -> handlePause(result)
      RESUME -> handleResume(result)
      STOP -> handleStop(result)
      else -> result.notImplemented()
    }
  }
  //endregion

  //region Recorder
  private fun handleInit(call: MethodCall, result: Result) {
//    TODO - CHRIS - need to handle the business logic for determining when the states can be changed
//    possibly return result.error when cannot init, with message like "already init and recording, must stop recording first"
    if (recorderState == UNSET || recorderState == INITIALIZED || recorderState == STOPPED) {
      resetRecorder()

      filepath = call.argument<Any?>(FILEPATH)?.toString()
      extension = call.argument<Any?>(EXTENSION)?.toString()
      sampleRate = call.argument<Any?>(SAMPLE_RATE)?.toString()?.toLong() ?: sampleRate
      bufferSize = AudioRecord.getMinBufferSize(sampleRate.toInt(), CHANNEL_IN_MONO, ENCODING_PCM_16BIT)
      recorderState = if (filepath.isNullOrBlank().not() && extension.isNullOrBlank().not()) INITIALIZED else UNSET
      message = ""
      result.success(recording)
    } else {
      result.error()
    }
  }

  private fun handleCurrent(result: Result) = result.success(recording)

  private fun handleStart(result: Result) {
    recorder = AudioRecord(MIC, sampleRate.toInt(), CHANNEL_IN_MONO, ENCODING_PCM_16BIT, bufferSize)

    try {
      fileOutputStream = FileOutputStream(filepathTemp)
    } catch (exception: FileNotFoundException) {
      result.error("", "cannot find the file", null)
      return
    }

    recorder?.startRecording()
    recorderState = RECORDING
    startThread()
    result.success(recording)
  }

  private fun handlePause(result: Result) {
    recorderState = PAUSED
    peakPower = DEFAULT_PEAK_POWER
    averagePower = DEFAULT_AVERAGE_POWER
    recorder?.stop()
    recordingThread = null
    result.success(recording)
  }

  private fun handleResume(result: Result) {
    recorderState = RECORDING
    recorder?.startRecording()
    startThread()
    result.success(recording)
  }

  private fun handleStop(result: Result) {
    if (recorderState == STOPPED) {
      result.success(recording)
    } else {
      recorderState = STOPPED

      resetRecorder()
      recordingThread = null
      recorder?.stop()
      recorder?.release()

      try {
        fileOutputStream?.close()
      } catch (exception: IOException) {
        exception.printStackTrace()
      }

      filepathTemp?.let { tempFileName ->
        filepath?.let { filePath ->
          copyWaveFile(tempFileName, filePath)
        }
      }

      deleteTempFile()

      result.success(recording)
    }
  }

  private fun resetRecorder() {
    peakPower = DEFAULT_PEAK_POWER
    averagePower = DEFAULT_AVERAGE_POWER
    dataSize = DEFAULT_DATA_SIZE
  }
  //endregion

  //region Audio Stream
  private fun processAudioStream() {
    val size = bufferSize
    val bData = ByteArray(size)

    while (recorderState === RECORDING) {
      recorder?.read(bData, 0, bData.size)
      dataSize += bData.size.toLong()
      updatePowers(bData)

      try {
        fileOutputStream?.write(bData)
      } catch (e: IOException) {
        e.printStackTrace()
      }
    }
  }

  private fun updatePowers(byteArray: ByteArray) {
    val shortArray = byteArray.toShortArray()
    val sampleVal: Short = shortArray.last()
    val escapeRecorderStateList = arrayOf(PAUSED, STOPPED, INITIALIZED, UNSET)

    averagePower = if (sampleVal == 0.toShort() || escapeRecorderStateList.contains(recorderState)) {
      DEFAULT_AVERAGE_POWER //to match iOS silent case
    } else {
      20 * ln(abs(sampleVal.toDouble()) / 32768.0) * IOS_POWER_LEVEL_FACTOR
    }

    peakPower = averagePower
  }
  //endregion

  //region Threads
  private fun startThread() {
    recordingThread = Thread({ processAudioStream() }, "Audio Processing Thread")
    recordingThread?.start()
  }
  //endregion

  //region Files
  private fun deleteTempFile() {
    filepathTemp?.let {
      val file = File(it)
      if (file.exists()) file.delete()
    }
  }

  private fun copyWaveFile(inputFileName: String, outputFileName: String) {
    try {
      val inputFileStream = FileInputStream(inputFileName)
      val outputFileStream = FileOutputStream(outputFileName)
      val totalAudioLength = inputFileStream.channel.size()
      val totalDataLength = totalAudioLength + 36

      val channels = 1
      val byteRate = RECORDER_BPP * sampleRate * channels / 8
      writeWaveFileHeader(outputFileStream, totalAudioLength, totalDataLength, sampleRate, channels, byteRate)

      val data = ByteArray(bufferSize)
      while (inputFileStream.read(data) != -1) {
        outputFileStream.write(data)
      }

      inputFileStream.close()
      outputFileStream.close()

    } catch (exception: FileNotFoundException) {
      exception.printStackTrace()
    } catch (exception: IOException) {
      exception.printStackTrace()
    }
  }

  @Throws (IOException::class)
  private fun writeWaveFileHeader(
    out: FileOutputStream,
    totalAudioLength: Long,
    totalDataLength: Long,
    longSampleRate: Long,
    channels: Int,
    byteRate: Long
  ) {
    try {
      val header = ByteArray(44)

      header[0] = 'R'.code.toByte() // RIFF/WAVE header

      header[1] = 'I'.code.toByte()
      header[2] = 'F'.code.toByte()
      header[3] = 'F'.code.toByte()
      header[4] = (totalDataLength and 0xffL).toByte()
      header[5] = (totalDataLength shr 8 and 0xffL).toByte()
      header[6] = (totalDataLength shr 16 and 0xffL).toByte()
      header[7] = (totalDataLength shr 24 and 0xffL).toByte()
      header[8] = 'W'.code.toByte()
      header[9] = 'A'.code.toByte()
      header[10] = 'V'.code.toByte()
      header[11] = 'E'.code.toByte()
      header[12] = 'f'.code.toByte() // 'fmt ' chunk

      header[13] = 'm'.code.toByte()
      header[14] = 't'.code.toByte()
      header[15] = ' '.code.toByte()
      header[16] = 16 // 4 bytes: size of 'fmt ' chunk

      header[17] = 0
      header[18] = 0
      header[19] = 0
      header[20] = 1 // format = 1

      header[21] = 0
      header[22] = channels.toByte()
      header[23] = 0
      header[24] = (longSampleRate and 0xffL).toByte()
      header[25] = (longSampleRate shr 8 and 0xffL).toByte()
      header[26] = (longSampleRate shr 16 and 0xffL).toByte()
      header[27] = (longSampleRate shr 24 and 0xffL).toByte()
      header[28] = (byteRate and 0xffL).toByte()
      header[29] = (byteRate shr 8 and 0xffL).toByte()
      header[30] = (byteRate shr 16 and 0xffL).toByte()
      header[31] = (byteRate shr 24 and 0xffL).toByte()
      header[32] = (channels * RECORDER_BPP.toInt() shr 3).toByte()
      header[33] = (channels * RECORDER_BPP.toInt() shr 11).toByte()
      header[34] = RECORDER_BPP // bits per sample

      header[35] = 0
      header[36] = 'd'.code.toByte()
      header[37] = 'a'.code.toByte()
      header[38] = 't'.code.toByte()
      header[39] = 'a'.code.toByte()
      header[40] = (totalAudioLength and 0xffL).toByte()
      header[41] = (totalAudioLength shr 8 and 0xffL).toByte()
      header[42] = (totalAudioLength shr 16 and 0xffL).toByte()
      header[43] = (totalAudioLength shr 24 and 0xffL).toByte()

      out.write(header, 0, 44)
    } catch (exception: IOException) {
      throw exception
    }
  }
  //endregion
}
